---
title: "celltalker"
author: "Anthony Cillo"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{celltalker}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r,echo=FALSE,message=FALSE,warning=FALSE}
library(knitr)
hook_output=knit_hooks$get('output')
knit_hooks$set(output=function(x,options) {
  if (is.null(n <- options$linewidth)) {
    x = knitr::split_lines(x)
    if (any(nchar(x)>n)) x = strwrap(x,width=n)
    x = paste(x,collapse = '\n')
  }
  hook_output(x,options)
})
```

```{r,include=FALSE}
  path.to.working <- "/Users/arc85/Desktop/celltalker_vignette.20.01.02/"
```

# Introduction

Amongst the many types of analysis possible with single-cell RNAseq data is the assessment of putative cell-cell communication. *celltalker* seeks to evaluate cell-cell communication (that is, "talking") by looking for expression of known pairs of ligands and receptors within and between cell populations. Any set of known ligands and receptors can be used; included in this package is the set of ligands and receptors described by the excellent work of Ramilowski et al (Nature Communications, 2015). We recommend using this list of ligands and receptors as a starting point, and either curating your own list of known ligands and receptors. Alternatively, work by Vento-Tormo et al (Nature, 2018) has recently extensively curated ligands and receptors, and will be implemented in future versions of cellTalker.

For the sake of robustness, we assume that there are multiple replicate samples in each group and that we are only interested in only finding consistently expressed ligands and receptors across groups (as opposed to potentially suprious interactions that are only found in one individual replicate). We accomplish this by looking at individual expression matrices for each individual sample in a group, and pulling only those interactions that meet a given (arbitrary) threshold.

In this vignette, we demonstrate a basic use of cellTalker to evaluate putative ligand/receptor interactions in peripheral blood (N=2) and tonsils from healthy donors (N=3). This data is publically available from our recently published dataset (Cillo et al, Immunity 2020) on the NCBI Gene Expression Omnibus Series GSE139324.


# Vignette overview

This vignette demonstrates a standard use of celltalker on publically available data from 10X Genomics. We will go through the following steps:

- Cluster our data using a standard Seurat workflow (v.3.1.1)
- Use celltalker to build lists of consistently expressed ligands and receptors within samples groups
- Determine putative ligand/receptor interactions with a groups
- Evaluate uniquely expressed ligand/receptor pairs between groups
- Identifying and visualizing unique ligand/receptor pairs in a group


# Clustering data with Seurat

Our vignette begins with a standard clustering analysis and identification of immune lineages in Seurat, and assumes that the feature/barcode matrices have been downloaded from GEO.

```{r,fig.width=7,fig.asp=0.618,out.width="70%",fig.align="center",linewidth=60}
suppressMessages({
library(Seurat)
library(celltalker)
})

#Set seed for reproducibility

set.seed(02221989)

#Read in raw data

setwd("data_matrices")
data.paths <- list.files()
specific.paths <- paste(path.to.working,"data_matrices",data.paths,"GRCh38",sep="/")
setwd(path.to.working)

raw.data <- Read10X(specific.paths)

#Create metadata

sample.data <- data.frame(matrix(data=NA,nrow=ncol(raw.data),ncol=2))
rownames(sample.data) <- colnames(raw.data)
colnames(sample.data) <- c("sample.id","sample.type")

sample.data[grep("^[A-z]",rownames(sample.data)),"sample.id"] <- "pbmc_1"
sample.data[grep("^2",rownames(sample.data)),"sample.id"] <- "tonsil_1"
sample.data[grep("^3",rownames(sample.data)),"sample.id"] <- "pbmc_2"
sample.data[grep("^4",rownames(sample.data)),"sample.id"] <- "tonsil_2"
sample.data[grep("^5",rownames(sample.data)),"sample.id"] <- "pbmc_3"
sample.data[grep("^6",rownames(sample.data)),"sample.id"] <- "tonsil_3"

sample.data[,"sample.type"] <- sapply(strsplit(sample.data$sample.id,split="_"),function(x) x[1])

#Create a Seurat object with associated metadata

ser.obj <- CreateSeuratObject(counts=raw.data,meta.data=sample.data)

#Standard Seurat workflow

ser.obj <- NormalizeData(ser.obj)
ser.obj <- FindVariableFeatures(ser.obj)
ser.obj <- ScaleData(ser.obj)
ser.obj <- RunPCA(ser.obj)

ElbowPlot(ser.obj)

#We will select the first 15 PCs to use

ser.obj <- RunUMAP(ser.obj,reduction="pca",dims=1:15)
ser.obj <- FindNeighbors(ser.obj,reduction="pca",dims=1:15)
ser.obj <- FindClusters(ser.obj,resolution=0.5)

```

Let's have a look at our data.

```{r,fig.width=7,fig.asp=0.618,fig.align="center",linewidth=60}

p1 <- DimPlot(ser.obj,reduction="umap",group.by="sample.id")
p2 <- DimPlot(ser.obj,reduction="umap",group.by="sample.type")
p3 <- DimPlot(ser.obj,reduction="umap",group.by="RNA_snn_res.0.5",label=T) + NoLegend()

cowplot::plot_grid(p1,p2,p3)

```

Great! Now let's (roughly) identify our cell types by having a look at a feature plot.

```{r,fig.width=7,fig.asp=0.618,fig.align="center",linewidth=60}

FeaturePlot(ser.obj,reduction="umap",features=c("CD3D","CD8A","CD4","CD14","MS4A1","FCGR3A","IL3RA"))

#Add metadata for cell types

cell.types <- vector("logical",length=ncol(ser.obj))
names(cell.types) <- colnames(ser.obj)

cell.types[ser.obj@meta.data$RNA_snn_res.0.5=="0"] <- "cd4.tconv"
cell.types[ser.obj@meta.data$RNA_snn_res.0.5=="1"] <- "cd4.tconv"
cell.types[ser.obj@meta.data$RNA_snn_res.0.5=="2"] <- "b.cells"
cell.types[ser.obj@meta.data$RNA_snn_res.0.5=="3"] <- "b.cells"
cell.types[ser.obj@meta.data$RNA_snn_res.0.5=="4"] <- "cd14.monocytes"
cell.types[ser.obj@meta.data$RNA_snn_res.0.5=="5"] <- "cd8.cells"
cell.types[ser.obj@meta.data$RNA_snn_res.0.5=="6"] <- "cd4.tconv"
cell.types[ser.obj@meta.data$RNA_snn_res.0.5=="7"] <- "cd4.tconv"
cell.types[ser.obj@meta.data$RNA_snn_res.0.5=="8"] <- "b.cells"
cell.types[ser.obj@meta.data$RNA_snn_res.0.5=="9"] <- "b.cells"
cell.types[ser.obj@meta.data$RNA_snn_res.0.5=="10"] <- "nk.cells"
cell.types[ser.obj@meta.data$RNA_snn_res.0.5=="11"] <- "cd8.cells"
cell.types[ser.obj@meta.data$RNA_snn_res.0.5=="12"] <- "plasma.cells"
cell.types[ser.obj@meta.data$RNA_snn_res.0.5=="13"] <- "cd14.monocytes"
cell.types[ser.obj@meta.data$RNA_snn_res.0.5=="14"] <- "cd16.monocytes"
cell.types[ser.obj@meta.data$RNA_snn_res.0.5=="15"] <- "pdc"
cell.types[ser.obj@meta.data$RNA_snn_res.0.5=="16"] <- "RBCs"

ser.obj[["cell.types"]] <- cell.types

#Let's remove RBCs from our analysis

rbc.cell.names <- names(cell.types)[ser.obj@meta.data$RNA_snn_res.0.5=="16"]

ser.obj <- ser.obj[,!colnames(ser.obj) %in% rbc.cell.names]

```

# celltalker
# Consistently expressed ligands and receptors

Now that we have identified and named clusters within our data, we will proceed with the celltalker analysis. Shipped with this packages is the data.frame "ramilowski_pairs", which is a data.frame consisting of ligands, receptors, and putative ligand_receptor pairs. Please refer to the "data-raw" folder, and the "create_ramolowski_pairs_data.R" file if your interested in how this data.frame was constructed.

First, we will perform differential gene expression analysis to
prune the universe of potential ligands and receptors down to those that are differentially expressed between our two sample groups. We identify ligands and receptors (from ramilowski_pairs) present in our overall dataset, then perform differential expression using those ligands and receptors.

Second, we will create a tibble with individual matrices for each replicate sample. We use tibbles (and more broadly, the tidyverse) throughout celltalker to make data manipulation more streamlined and to use the hierarchical structure of tibbles as containers for relevant pieces of data.

```{r,linewidth=60}

#Check out ramilowski_pairs data.frame

head(ramilowski_pairs)
dim(ramilowski_pairs)

#There are 2,557 unique ligand/receptor interactions in this dataset

#Identification of differentially expressed ligands and receptors

#Identify ligands and receptors in our dataset

ligs <- as.character(unique(ramilowski_pairs$ligand))
recs <- as.character(unique(ramilowski_pairs$receptor))

ligs.present <- rownames(ser.obj)[rownames(ser.obj) %in% ligs]
recs.present <- rownames(ser.obj)[rownames(ser.obj) %in% recs]

genes.to.use <- union(ligs.present,recs.present)

#Use FindAllMarkers for differentially expressed ligands and receptors between groups

Idents(ser.obj) <- "sample.type"
markers <- FindAllMarkers(ser.obj,assay="RNA",features=genes.to.use,only.pos=TRUE)

ligs.recs.use <- unique(markers$gene)
length(ligs.recs.use)
#Yields 61 ligands and receptors to evaluate

#Filter ramilowski pairs

interactions.forward1 <- ramilowski_pairs[as.character(ramilowski_pairs$ligand) %in% ligs.recs.use,]
interactions.forward2 <- ramilowski_pairs[as.character(ramilowski_pairs$receptor) %in% ligs.recs.use,]
interact.for <- rbind(interactions.forward1,interactions.forward2)
dim(interact.for)
#Yields 241 ligand and receptor interactions to evaluate

#Create data for celltalker

expr.mat <- GetAssayData(ser.obj,slot="counts")
defined.clusters <- ser.obj@meta.data$cell.types
defined.groups <- ser.obj@meta.data$sample.type
defined.replicates <- ser.obj@meta.data$sample.id

reshaped.matrices <- reshape_matrices(count.matrix=expr.mat,clusters=defined.clusters,groups=defined.groups,replicates=defined.replicates,ligands.and.receptors=interact.for)

#Check out the hierarchy of the tibble
reshaped.matrices
unnest(reshaped.matrices,cols="samples")
names(pull(unnest(reshaped.matrices,cols="samples"))[[1]])
```

What we have done in this initial step is to separate our overall expression matrix into a series of individual expression matrices from each sample. An overview of the resulting tibble is as follows:

2-column tibble containing group and samples
-  "samples" column from the top-level contains 3-column tibble
  - "expr.matrices" from above is a list of count matrices from each cluster of an individual replicate (in this case, clusters are the cell types)

Next, we can create a tibble of consistently expressed ligands and receptors for each group using the create_lig_rec_tib function.

```{r,linewidth=60}

consistent.lig.recs <- create_lig_rec_tib(exp.tib=reshaped.matrices,clusters=defined.clusters,groups=defined.groups,replicates=defined.replicates,cells.reqd=10,freq.pos.reqd=0.5,ligands.and.receptors=interact.for)

consistent.lig.recs
unnest(consistent.lig.recs[1,2],cols="lig.rec.exp")
pull(unnest(consistent.lig.recs[1,2],cols="lig.rec.exp")[1,2])[[1]]
```

We can see that we have now recovered a list of consistently expressed ligands and receptors from each cluster in each experimental group, based on the criteria specified above (i.e. at least 10 cells in each cluster express the ligand/receptor and at least 50% of replicates in a group express the ligand/receptor). Functionally, since we only have 2 replicates in each group here, we require all replicate to express the ligand/receptor to count it as consistently expressed.

# Determine putative ligand/receptor pairs

Now that we have consistently expressed ligands and receptors, we will look for interactions (as defined in ramilowski_pairs$pair) within and between clusters in a given group. The parameter "freq.group.in.cluster" allows us to only examine clusters that consist of (in this case) >5% of cells from a given group for interactions.

```{r,linewidth=60}

put.int <- putative_interactions(ligand.receptor.tibble=consistent.lig.recs,clusters=defined.clusters,groups=defined.groups,freq.group.in.cluster=0.05,ligands.and.receptors=interact.for)

```

Here, the nested tibble contains the sample groups and a list of ligand/receptor pairs for each group along with the clusters that particpate in the ligand/receptor interactions.

# Identifying and visualizing unique ligand/receptor pairs in a group

Now that we have our lists of ligand/receptor interactions, we can investigate which one are unique to each group (using the unique_interactions function) and plot the results (using the circos_plot function) to visualize the differences between groups.

```{r,fig.width=10,fig.asp=0.618,fig.align="center",linewidth=60}

#Identify unique ligand/receptor interactions present in each sample
unique.ints <- unique_interactions(put.int,group1="pbmc",group2="tonsil",interact.for)

#Get data to plot circos for PBMC
pbmc.to.plot <- pull(unique.ints[1,2])[[1]]
for.circos.pbmc <- pull(put.int[1,2])[[1]][pbmc.to.plot]

circos_plot(interactions=for.circos.pbmc,clusters=defined.clusters)

#Get data to plot circos for tonsil
tonsil.to.plot <- pull(unique.ints[2,2])[[1]]
for.circos.tonsil <- pull(put.int[2,2])[[1]][tonsil.to.plot]

circos_plot(interactions=for.circos.tonsil,clusters=defined.clusters)

```

# Summary
celltalker allows a user to infer ligand/receptor interactions. This vignette has walked through how to identify and visualize unique interactions, staring with raw count data and metadata. Future developments will focus on differential networks of cell-cell communication for given ligand/receptor pairs. Stay tuned for future updates. Until then, have fun investigating how your cells "talk" to each other!
