<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>celltalker • celltalker</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="celltalker">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">celltalker</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.3.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/celltalker.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="celltalker_files/jquery-1.11.3/jquery.min.js"></script><script src="celltalker_files/elevate-section-attrs-2.0/elevate-section-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>celltalker</h1>
                        <h4 class="author">Anthony Cillo</h4>
            
            <h4 class="date">2020-01-08</h4>
      
      
      <div class="hidden name"><code>celltalker.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>Amongst the many types of analysis possible with single-cell RNAseq data is the assessment of putative cell-cell communication. <em>celltalker</em> seeks to evaluate cell-cell communication (that is, “talking”) by looking for expression of known pairs of ligands and receptors within and between cell populations. Any set of known ligands and receptors can be used; included in this package is the set of ligands and receptors described by the excellent work of Ramilowski et al (Nature Communications, 2015). We recommend using this list of ligands and receptors as a starting point, and either curating your own list of known ligands and receptors. Alternatively, work by Vento-Tormo et al (Nature, 2018) has recently extensively curated ligands and receptors, and will be implemented in future versions of cellTalker.</p>
<p>For the sake of robustness, we assume that there are multiple replicate samples in each group and that we are only interested in only finding consistently expressed ligands and receptors across groups (as opposed to potentially suprious interactions that are only found in one individual replicate). We accomplish this by looking at individual expression matrices for each individual sample in a group, and pulling only those interactions that meet a given (arbitrary) threshold.</p>
<p>In this vignette, we demonstrate a basic use of cellTalker to evaluate putative ligand/receptor interactions in peripheral blood (N=2) and tonsils from healthy donors (N=3). This data is publically available from our recently published dataset (Cillo et al, Immunity 2020) on the NCBI Gene Expression Omnibus Series GSE139324.</p>
</div>
<div id="vignette-overview" class="section level1">
<h1 class="hasAnchor">
<a href="#vignette-overview" class="anchor"></a>Vignette overview</h1>
<p>This vignette demonstrates a standard use of celltalker on publically available data from 10X Genomics. We will go through the following steps:</p>
<ul>
<li>Cluster our data using a standard Seurat workflow (v.3.1.1)</li>
<li>Use celltalker to build lists of consistently expressed ligands and receptors within samples groups</li>
<li>Determine putative ligand/receptor interactions with a groups</li>
<li>Evaluate uniquely expressed ligand/receptor pairs between groups</li>
<li>Identifying and visualizing unique ligand/receptor pairs in a group</li>
</ul>
</div>
<div id="clustering-data-with-seurat" class="section level1">
<h1 class="hasAnchor">
<a href="#clustering-data-with-seurat" class="anchor"></a>Clustering data with Seurat</h1>
<p>Our vignette begins with a standard clustering analysis and identification of immune lineages in Seurat, and assumes that the feature/barcode matrices have been downloaded from GEO.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="kw"><a href="https://rdrr.io/r/base/message.html">suppressMessages</a></span>({</span>
<span id="cb1-2"><a href="#cb1-2"></a><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(Seurat)</span>
<span id="cb1-3"><a href="#cb1-3"></a><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(celltalker)</span>
<span id="cb1-4"><a href="#cb1-4"></a>})</span>
<span id="cb1-5"><a href="#cb1-5"></a></span>
<span id="cb1-6"><a href="#cb1-6"></a><span class="co">#Set seed for reproducibility</span></span>
<span id="cb1-7"><a href="#cb1-7"></a></span>
<span id="cb1-8"><a href="#cb1-8"></a><span class="kw"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="dv">02221989</span>)</span>
<span id="cb1-9"><a href="#cb1-9"></a></span>
<span id="cb1-10"><a href="#cb1-10"></a><span class="co">#Read in raw data</span></span>
<span id="cb1-11"><a href="#cb1-11"></a></span>
<span id="cb1-12"><a href="#cb1-12"></a><span class="kw"><a href="https://rdrr.io/r/base/getwd.html">setwd</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(path.to.working,<span class="st">"/data_matrices/"</span>,<span class="dt">sep=</span><span class="st">""</span>))</span>
<span id="cb1-13"><a href="#cb1-13"></a>data.paths &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span>()</span>
<span id="cb1-14"><a href="#cb1-14"></a>specific.paths &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(path.to.working,<span class="st">"data_matrices"</span>,data.paths,<span class="st">"GRCh38"</span>,<span class="dt">sep=</span><span class="st">"/"</span>)</span>
<span id="cb1-15"><a href="#cb1-15"></a><span class="kw"><a href="https://rdrr.io/r/base/getwd.html">setwd</a></span>(path.to.working)</span>
<span id="cb1-16"><a href="#cb1-16"></a></span>
<span id="cb1-17"><a href="#cb1-17"></a>raw.data &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/Seurat/man/Read10X.html">Read10X</a></span>(specific.paths)</span>
<span id="cb1-18"><a href="#cb1-18"></a></span>
<span id="cb1-19"><a href="#cb1-19"></a><span class="co">#Create metadata</span></span>
<span id="cb1-20"><a href="#cb1-20"></a></span>
<span id="cb1-21"><a href="#cb1-21"></a>sample.data &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span>(<span class="dt">data=</span><span class="ot">NA</span>,<span class="dt">nrow=</span><span class="kw"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span>(raw.data),<span class="dt">ncol=</span><span class="dv">2</span>))</span>
<span id="cb1-22"><a href="#cb1-22"></a><span class="kw"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(sample.data) &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(raw.data)</span>
<span id="cb1-23"><a href="#cb1-23"></a><span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(sample.data) &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"sample.id"</span>,<span class="st">"sample.type"</span>)</span>
<span id="cb1-24"><a href="#cb1-24"></a></span>
<span id="cb1-25"><a href="#cb1-25"></a>sample.data[<span class="kw"><a href="https://rdrr.io/r/base/grep.html">grep</a></span>(<span class="st">"^[A-z]"</span>,<span class="kw"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(sample.data)),<span class="st">"sample.id"</span>] &lt;-<span class="st"> "pbmc_1"</span></span>
<span id="cb1-26"><a href="#cb1-26"></a>sample.data[<span class="kw"><a href="https://rdrr.io/r/base/grep.html">grep</a></span>(<span class="st">"^2"</span>,<span class="kw"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(sample.data)),<span class="st">"sample.id"</span>] &lt;-<span class="st"> "tonsil_1"</span></span>
<span id="cb1-27"><a href="#cb1-27"></a>sample.data[<span class="kw"><a href="https://rdrr.io/r/base/grep.html">grep</a></span>(<span class="st">"^3"</span>,<span class="kw"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(sample.data)),<span class="st">"sample.id"</span>] &lt;-<span class="st"> "pbmc_2"</span></span>
<span id="cb1-28"><a href="#cb1-28"></a>sample.data[<span class="kw"><a href="https://rdrr.io/r/base/grep.html">grep</a></span>(<span class="st">"^4"</span>,<span class="kw"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(sample.data)),<span class="st">"sample.id"</span>] &lt;-<span class="st"> "tonsil_2"</span></span>
<span id="cb1-29"><a href="#cb1-29"></a>sample.data[<span class="kw"><a href="https://rdrr.io/r/base/grep.html">grep</a></span>(<span class="st">"^5"</span>,<span class="kw"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(sample.data)),<span class="st">"sample.id"</span>] &lt;-<span class="st"> "pbmc_3"</span></span>
<span id="cb1-30"><a href="#cb1-30"></a>sample.data[<span class="kw"><a href="https://rdrr.io/r/base/grep.html">grep</a></span>(<span class="st">"^6"</span>,<span class="kw"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(sample.data)),<span class="st">"sample.id"</span>] &lt;-<span class="st"> "tonsil_3"</span></span>
<span id="cb1-31"><a href="#cb1-31"></a></span>
<span id="cb1-32"><a href="#cb1-32"></a>sample.data[,<span class="st">"sample.type"</span>] &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/strsplit.html">strsplit</a></span>(sample.data<span class="op">$</span>sample.id,<span class="dt">split=</span><span class="st">"_"</span>),<span class="cf">function</span>(x) x[<span class="dv">1</span>])</span>
<span id="cb1-33"><a href="#cb1-33"></a></span>
<span id="cb1-34"><a href="#cb1-34"></a><span class="co">#Create a Seurat object with associated metadata</span></span>
<span id="cb1-35"><a href="#cb1-35"></a></span>
<span id="cb1-36"><a href="#cb1-36"></a>ser.obj &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/Seurat/man/CreateSeuratObject.html">CreateSeuratObject</a></span>(<span class="dt">counts=</span>raw.data,<span class="dt">meta.data=</span>sample.data)</span></code></pre></div>
<pre><code>## Warning: Feature names cannot have underscores ('_'), replacing with dashes
## ('-')</code></pre>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a><span class="co">#Standard Seurat workflow</span></span>
<span id="cb3-2"><a href="#cb3-2"></a></span>
<span id="cb3-3"><a href="#cb3-3"></a>ser.obj &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/Seurat/man/NormalizeData.html">NormalizeData</a></span>(ser.obj)</span>
<span id="cb3-4"><a href="#cb3-4"></a>ser.obj &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/Seurat/man/FindVariableFeatures.html">FindVariableFeatures</a></span>(ser.obj)</span>
<span id="cb3-5"><a href="#cb3-5"></a>ser.obj &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/Seurat/man/ScaleData.html">ScaleData</a></span>(ser.obj)</span></code></pre></div>
<pre><code>## Centering and scaling data matrix</code></pre>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a>ser.obj &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/Seurat/man/RunPCA.html">RunPCA</a></span>(ser.obj)</span></code></pre></div>
<pre><code>## PC_ 1 
## Positive:  S100A6, IL32, S100A4, ANXA1, VIM, FTL, TRBC1, SRGN, S100A9, S100A8 
##     TYROBP, LYZ, CTSW, XIST, NEAT1, VCAN, S100A12, FCER1G, S100A11, FCN1 
##     PLAC8, ID2, CCL5, NKG7, CST3, CSTA, ZFP36, IL1B, MT2A, KLRB1 
## Negative:  RGS13, KIAA0101, NUSAP1, AURKB, MKI67, BIRC5, TYMS, TOP2A, TK1, CDKN3 
##     UBE2C, PTTG1, CDK1, STMN1, CCNB2, GTSE1, BIK, RRM2, TCL1A, SHCBP1 
##     CDCA3, CDC20, TPX2, LRMP, CCNA2, MND1, CCNB1, PBK, ZWINT, RMI2 
## PC_ 2 
## Positive:  CST3, LYZ, FCN1, CSTA, S100A9, S100A8, TYROBP, LST1, FGL2, VCAN 
##     S100A12, SERPINA1, MNDA, FCER1G, CLEC7A, MS4A6A, CD14, CFD, IL1B, TYMP 
##     LGALS1, RP11-1143G9.4, AIF1, CTSS, NAMPT, CFP, TNFSF13B, CSF3R, MPEG1, TMEM176B 
## Negative:  IL32, NPM1, CD69, TRBC1, ISG20, ITM2A, IGKC, IGHA1, HSP90AB1, DDIT4 
##     HIST1H4C, PSIP1, AQP3, MYC, PIM2, HMGN1, PASK, NUCB2, HSPA1B, HSPB1 
##     CD79A, SUSD3, KLRB1, SYNE2, CHI3L2, IGHG3, IGLC2, FKBP11, IGHG1, SH2D1A 
## PC_ 3 
## Positive:  IL32, NKG7, CTSW, TRBC1, GZMA, CST7, GNLY, MKI67, ANXA1, TOP2A 
##     CCL5, PRF1, BIRC5, S100A4, KLRB1, CCNA2, AURKB, CENPF, GTSE1, CDKN3 
##     KLRD1, UBE2C, CDK1, TYMS, TPX2, RRM2, ID2, S100A6, FGFBP2, CDC20 
## Negative:  HLA-DRA, HLA-DQA1, HLA-DQB1, CD79A, HLA-DRB1, MS4A1, CD74, HLA-DPA1, HLA-DPB1, CD79B 
##     HLA-DMA, HLA-DMB, BANK1, VPREB3, IGKC, HLA-DRB5, MEF2C, CD22, IRF8, CD19 
##     SMIM14, FCRLA, HLA-DOB, CD24, CD40, FCER2, BLK, HLA-DQA2, IGHD, CTSH 
## PC_ 4 
## Positive:  TOP2A, UBE2C, MKI67, GTSE1, CENPF, AURKB, PLK1, CCNA2, CDK1, CDCA8 
##     HMMR, CDCA3, CDC20, TPX2, CDKN3, DLGAP5, CENPE, BIRC5, CCNB2, CENPA 
##     KIF2C, CKAP2L, PBK, NUSAP1, KIFC1, AURKA, SPC25, NUF2, KIF23, ASPM 
## Negative:  NKG7, GNLY, CST7, GZMB, GZMA, PRF1, KLRD1, FGFBP2, CCL5, KLRF1 
##     HOPX, CTSW, GZMH, TRDC, FCGR3A, SPON2, CLIC3, MATK, ADGRG1, S1PR5 
##     CCL4, CMC1, XCL2, PFN1, CD160, FCRL6, IL2RB, TRGC1, KLRC1, C12orf75 
## PC_ 5 
## Positive:  ICA1, PDCD1, TBC1D4, ITM2A, ICOS, MAF, TOX2, IL32, TNFRSF4, PASK 
##     PKM, SMCO4, ACTG1, CORO1B, CTLA4, NPM1, TRBC1, PCAT29, TIGIT, AC133644.2 
##     TOX, ANP32B, ENO1, GBP2, COTL1, GAPDH, SUSD3, PIM2, AQP3, SERPINA9 
## Negative:  NKG7, GNLY, KLRD1, FGFBP2, GZMB, GZMA, KLRF1, CCL5, PRF1, TRDC 
##     GZMH, CST7, CTSW, BANK1, MATK, PLK1, HMMR, HLA-DPB1, CENPA, CLIC3 
##     GTSE1, CENPE, CCL4, SPON2, PDLIM1, HLA-DPA1, CDCA8, DLGAP5, TPX2, IGHD</code></pre>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a><span class="kw"><a href="https://rdrr.io/pkg/Seurat/man/ElbowPlot.html">ElbowPlot</a></span>(ser.obj)</span></code></pre></div>
<p><img src="celltalker_files/figure-html/unnamed-chunk-3-1.png" width="70%" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a><span class="co">#We will select the first 15 PCs to use</span></span>
<span id="cb8-2"><a href="#cb8-2"></a></span>
<span id="cb8-3"><a href="#cb8-3"></a>ser.obj &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/Seurat/man/RunUMAP.html">RunUMAP</a></span>(ser.obj,<span class="dt">reduction=</span><span class="st">"pca"</span>,<span class="dt">dims=</span><span class="dv">1</span><span class="op">:</span><span class="dv">15</span>)</span></code></pre></div>
<pre><code>## Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric
## To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'
## This message will be shown once per session</code></pre>
<pre><code>## 00:25:40 UMAP embedding parameters a = 0.9922 b = 1.112</code></pre>
<pre><code>## 00:25:40 Read 15524 rows and found 15 numeric columns</code></pre>
<pre><code>## 00:25:40 Using Annoy for neighbor search, n_neighbors = 30</code></pre>
<pre><code>## 00:25:40 Building Annoy index with metric = cosine, n_trees = 50</code></pre>
<pre><code>## 0%   10   20   30   40   50   60   70   80   90   100%</code></pre>
<pre><code>## [----|----|----|----|----|----|----|----|----|----|</code></pre>
<pre><code>## **************************************************|
## 00:25:42 Writing NN index file to temp file /var/folders/_p/n2bykf5555bcpmdl3yfw8y540000gn/T//Rtmp1P2ywh/file180f66b002cad
## 00:25:42 Searching Annoy index using 1 thread, search_k = 3000
## 00:25:47 Annoy recall = 100%
## 00:25:48 Commencing smooth kNN distance calibration using 1 thread
## 00:25:48 Initializing from normalized Laplacian + noise
## 00:25:50 Commencing optimization for 200 epochs, with 656278 positive edges
## 00:26:00 Optimization finished</code></pre>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1"></a>ser.obj &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/Seurat/man/FindNeighbors.html">FindNeighbors</a></span>(ser.obj,<span class="dt">reduction=</span><span class="st">"pca"</span>,<span class="dt">dims=</span><span class="dv">1</span><span class="op">:</span><span class="dv">15</span>)</span></code></pre></div>
<pre><code>## Computing nearest neighbor graph
## Computing SNN</code></pre>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1"></a>ser.obj &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/Seurat/man/FindClusters.html">FindClusters</a></span>(ser.obj,<span class="dt">resolution=</span><span class="fl">0.5</span>)</span></code></pre></div>
<pre><code>## Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck
## 
## Number of nodes: 15524
## Number of edges: 543084
## 
## Running Louvain algorithm...
## Maximum modularity in 10 random starts: 0.9185
## Number of communities: 17
## Elapsed time: 2 seconds</code></pre>
<p>Let’s have a look at our data.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1"></a>p1 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/Seurat/man/DimPlot.html">DimPlot</a></span>(ser.obj,<span class="dt">reduction=</span><span class="st">"umap"</span>,<span class="dt">group.by=</span><span class="st">"sample.id"</span>)</span>
<span id="cb21-2"><a href="#cb21-2"></a>p2 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/Seurat/man/DimPlot.html">DimPlot</a></span>(ser.obj,<span class="dt">reduction=</span><span class="st">"umap"</span>,<span class="dt">group.by=</span><span class="st">"sample.type"</span>)</span>
<span id="cb21-3"><a href="#cb21-3"></a>p3 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/Seurat/man/DimPlot.html">DimPlot</a></span>(ser.obj,<span class="dt">reduction=</span><span class="st">"umap"</span>,<span class="dt">group.by=</span><span class="st">"RNA_snn_res.0.5"</span>,<span class="dt">label=</span>T) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/Seurat/man/SeuratTheme.html">NoLegend</a></span>()</span></code></pre></div>
<pre><code>## Warning: Using `as.character()` on a quosure is deprecated as of rlang 0.3.0.
## Please use `as_label()` or `as_name()` instead.
## This warning is displayed once per session.</code></pre>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1"></a>cowplot<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/cowplot/man/plot_grid.html">plot_grid</a></span>(p1,p2,p3)</span></code></pre></div>
<p><img src="celltalker_files/figure-html/unnamed-chunk-4-1.png" width="672" style="display: block; margin: auto;"></p>
<p>Great! Now let’s (roughly) identify our cell types by having a look at a feature plot.</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1"></a><span class="kw"><a href="https://rdrr.io/pkg/Seurat/man/FeaturePlot.html">FeaturePlot</a></span>(ser.obj,<span class="dt">reduction=</span><span class="st">"umap"</span>,<span class="dt">features=</span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"CD3D"</span>,<span class="st">"CD8A"</span>,<span class="st">"CD4"</span>,<span class="st">"CD14"</span>,<span class="st">"MS4A1"</span>,<span class="st">"FCGR3A"</span>,<span class="st">"IL3RA"</span>))</span></code></pre></div>
<p><img src="celltalker_files/figure-html/unnamed-chunk-5-1.png" width="672" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1"></a><span class="co">#Add metadata for cell types</span></span>
<span id="cb25-2"><a href="#cb25-2"></a></span>
<span id="cb25-3"><a href="#cb25-3"></a>cell.types &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/vector.html">vector</a></span>(<span class="st">"logical"</span>,<span class="dt">length=</span><span class="kw"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span>(ser.obj))</span>
<span id="cb25-4"><a href="#cb25-4"></a><span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(cell.types) &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(ser.obj)</span>
<span id="cb25-5"><a href="#cb25-5"></a></span>
<span id="cb25-6"><a href="#cb25-6"></a>cell.types[ser.obj<span class="op">@</span>meta.data<span class="op">$</span>RNA_snn_res.<span class="fl">0.5</span><span class="op">==</span><span class="st">"0"</span>] &lt;-<span class="st"> "cd4.tconv"</span></span>
<span id="cb25-7"><a href="#cb25-7"></a>cell.types[ser.obj<span class="op">@</span>meta.data<span class="op">$</span>RNA_snn_res.<span class="fl">0.5</span><span class="op">==</span><span class="st">"1"</span>] &lt;-<span class="st"> "cd4.tconv"</span></span>
<span id="cb25-8"><a href="#cb25-8"></a>cell.types[ser.obj<span class="op">@</span>meta.data<span class="op">$</span>RNA_snn_res.<span class="fl">0.5</span><span class="op">==</span><span class="st">"2"</span>] &lt;-<span class="st"> "b.cells"</span></span>
<span id="cb25-9"><a href="#cb25-9"></a>cell.types[ser.obj<span class="op">@</span>meta.data<span class="op">$</span>RNA_snn_res.<span class="fl">0.5</span><span class="op">==</span><span class="st">"3"</span>] &lt;-<span class="st"> "b.cells"</span></span>
<span id="cb25-10"><a href="#cb25-10"></a>cell.types[ser.obj<span class="op">@</span>meta.data<span class="op">$</span>RNA_snn_res.<span class="fl">0.5</span><span class="op">==</span><span class="st">"4"</span>] &lt;-<span class="st"> "cd14.monocytes"</span></span>
<span id="cb25-11"><a href="#cb25-11"></a>cell.types[ser.obj<span class="op">@</span>meta.data<span class="op">$</span>RNA_snn_res.<span class="fl">0.5</span><span class="op">==</span><span class="st">"5"</span>] &lt;-<span class="st"> "cd8.cells"</span></span>
<span id="cb25-12"><a href="#cb25-12"></a>cell.types[ser.obj<span class="op">@</span>meta.data<span class="op">$</span>RNA_snn_res.<span class="fl">0.5</span><span class="op">==</span><span class="st">"6"</span>] &lt;-<span class="st"> "cd4.tconv"</span></span>
<span id="cb25-13"><a href="#cb25-13"></a>cell.types[ser.obj<span class="op">@</span>meta.data<span class="op">$</span>RNA_snn_res.<span class="fl">0.5</span><span class="op">==</span><span class="st">"7"</span>] &lt;-<span class="st"> "cd4.tconv"</span></span>
<span id="cb25-14"><a href="#cb25-14"></a>cell.types[ser.obj<span class="op">@</span>meta.data<span class="op">$</span>RNA_snn_res.<span class="fl">0.5</span><span class="op">==</span><span class="st">"8"</span>] &lt;-<span class="st"> "b.cells"</span></span>
<span id="cb25-15"><a href="#cb25-15"></a>cell.types[ser.obj<span class="op">@</span>meta.data<span class="op">$</span>RNA_snn_res.<span class="fl">0.5</span><span class="op">==</span><span class="st">"9"</span>] &lt;-<span class="st"> "b.cells"</span></span>
<span id="cb25-16"><a href="#cb25-16"></a>cell.types[ser.obj<span class="op">@</span>meta.data<span class="op">$</span>RNA_snn_res.<span class="fl">0.5</span><span class="op">==</span><span class="st">"10"</span>] &lt;-<span class="st"> "nk.cells"</span></span>
<span id="cb25-17"><a href="#cb25-17"></a>cell.types[ser.obj<span class="op">@</span>meta.data<span class="op">$</span>RNA_snn_res.<span class="fl">0.5</span><span class="op">==</span><span class="st">"11"</span>] &lt;-<span class="st"> "cd8.cells"</span></span>
<span id="cb25-18"><a href="#cb25-18"></a>cell.types[ser.obj<span class="op">@</span>meta.data<span class="op">$</span>RNA_snn_res.<span class="fl">0.5</span><span class="op">==</span><span class="st">"12"</span>] &lt;-<span class="st"> "plasma.cells"</span></span>
<span id="cb25-19"><a href="#cb25-19"></a>cell.types[ser.obj<span class="op">@</span>meta.data<span class="op">$</span>RNA_snn_res.<span class="fl">0.5</span><span class="op">==</span><span class="st">"13"</span>] &lt;-<span class="st"> "cd14.monocytes"</span></span>
<span id="cb25-20"><a href="#cb25-20"></a>cell.types[ser.obj<span class="op">@</span>meta.data<span class="op">$</span>RNA_snn_res.<span class="fl">0.5</span><span class="op">==</span><span class="st">"14"</span>] &lt;-<span class="st"> "cd16.monocytes"</span></span>
<span id="cb25-21"><a href="#cb25-21"></a>cell.types[ser.obj<span class="op">@</span>meta.data<span class="op">$</span>RNA_snn_res.<span class="fl">0.5</span><span class="op">==</span><span class="st">"15"</span>] &lt;-<span class="st"> "pdc"</span></span>
<span id="cb25-22"><a href="#cb25-22"></a>cell.types[ser.obj<span class="op">@</span>meta.data<span class="op">$</span>RNA_snn_res.<span class="fl">0.5</span><span class="op">==</span><span class="st">"16"</span>] &lt;-<span class="st"> "RBCs"</span></span>
<span id="cb25-23"><a href="#cb25-23"></a></span>
<span id="cb25-24"><a href="#cb25-24"></a>ser.obj[[<span class="st">"cell.types"</span>]] &lt;-<span class="st"> </span>cell.types</span>
<span id="cb25-25"><a href="#cb25-25"></a></span>
<span id="cb25-26"><a href="#cb25-26"></a><span class="co">#Let's remove RBCs from our analysis</span></span>
<span id="cb25-27"><a href="#cb25-27"></a></span>
<span id="cb25-28"><a href="#cb25-28"></a>rbc.cell.names &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(cell.types)[ser.obj<span class="op">@</span>meta.data<span class="op">$</span>RNA_snn_res.<span class="fl">0.5</span><span class="op">==</span><span class="st">"16"</span>]</span>
<span id="cb25-29"><a href="#cb25-29"></a></span>
<span id="cb25-30"><a href="#cb25-30"></a>ser.obj &lt;-<span class="st"> </span>ser.obj[,<span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(ser.obj) <span class="op">%in%</span><span class="st"> </span>rbc.cell.names]</span></code></pre></div>
</div>
<div id="celltalker" class="section level1">
<h1 class="hasAnchor">
<a href="#celltalker" class="anchor"></a>celltalker</h1>
</div>
<div id="consistently-expressed-ligands-and-receptors" class="section level1">
<h1 class="hasAnchor">
<a href="#consistently-expressed-ligands-and-receptors" class="anchor"></a>Consistently expressed ligands and receptors</h1>
<p>Now that we have identified and named clusters within our data, we will proceed with the celltalker analysis. Shipped with this packages is the data.frame “ramilowski_pairs”, which is a data.frame consisting of ligands, receptors, and putative ligand_receptor pairs. Please refer to the “data-raw” folder, and the “create_ramolowski_pairs_data.R” file if your interested in how this data.frame was constructed.</p>
<p>First, we will perform differential gene expression analysis to prune the universe of potential ligands and receptors down to those that are differentially expressed between our two sample groups. We identify ligands and receptors (from ramilowski_pairs) present in our overall dataset, then perform differential expression using those ligands and receptors.</p>
<p>Second, we will create a tibble with individual matrices for each replicate sample. We use tibbles (and more broadly, the tidyverse) throughout celltalker to make data manipulation more streamlined and to use the hierarchical structure of tibbles as containers for relevant pieces of data.</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1"></a><span class="co">#Check out ramilowski_pairs data.frame</span></span>
<span id="cb26-2"><a href="#cb26-2"></a></span>
<span id="cb26-3"><a href="#cb26-3"></a><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(ramilowski_pairs)</span></code></pre></div>
<pre><code>##   ligand receptor         pair
## 1    A2M     LRP1     A2M_LRP1
## 2  AANAT   MTNR1A AANAT_MTNR1A
## 3  AANAT   MTNR1B AANAT_MTNR1B
## 4    ACE    AGTR2    ACE_AGTR2
## 5    ACE   BDKRB2   ACE_BDKRB2
## 6 ADAM10      AXL   ADAM10_AXL</code></pre>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1"></a><span class="kw"><a href="https://rdrr.io/r/base/dim.html">dim</a></span>(ramilowski_pairs)</span></code></pre></div>
<pre><code>## [1] 2557    3</code></pre>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1"></a><span class="co">#There are 2,557 unique ligand/receptor interactions in this dataset</span></span>
<span id="cb30-2"><a href="#cb30-2"></a></span>
<span id="cb30-3"><a href="#cb30-3"></a><span class="co">#Identification of differentially expressed ligands and receptors</span></span>
<span id="cb30-4"><a href="#cb30-4"></a></span>
<span id="cb30-5"><a href="#cb30-5"></a><span class="co">#Identify ligands and receptors in our dataset</span></span>
<span id="cb30-6"><a href="#cb30-6"></a></span>
<span id="cb30-7"><a href="#cb30-7"></a>ligs &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/character.html">as.character</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(ramilowski_pairs<span class="op">$</span>ligand))</span>
<span id="cb30-8"><a href="#cb30-8"></a>recs &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/character.html">as.character</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(ramilowski_pairs<span class="op">$</span>receptor))</span>
<span id="cb30-9"><a href="#cb30-9"></a></span>
<span id="cb30-10"><a href="#cb30-10"></a>ligs.present &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(ser.obj)[<span class="kw"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(ser.obj) <span class="op">%in%</span><span class="st"> </span>ligs]</span>
<span id="cb30-11"><a href="#cb30-11"></a>recs.present &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(ser.obj)[<span class="kw"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(ser.obj) <span class="op">%in%</span><span class="st"> </span>recs]</span>
<span id="cb30-12"><a href="#cb30-12"></a></span>
<span id="cb30-13"><a href="#cb30-13"></a>genes.to.use &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/sets.html">union</a></span>(ligs.present,recs.present)</span>
<span id="cb30-14"><a href="#cb30-14"></a></span>
<span id="cb30-15"><a href="#cb30-15"></a><span class="co">#Use FindAllMarkers for differentially expressed ligands and receptors between groups</span></span>
<span id="cb30-16"><a href="#cb30-16"></a></span>
<span id="cb30-17"><a href="#cb30-17"></a><span class="kw"><a href="https://rdrr.io/pkg/Seurat/man/Idents.html">Idents</a></span>(ser.obj) &lt;-<span class="st"> "sample.type"</span></span>
<span id="cb30-18"><a href="#cb30-18"></a>markers &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/Seurat/man/FindAllMarkers.html">FindAllMarkers</a></span>(ser.obj,<span class="dt">assay=</span><span class="st">"RNA"</span>,<span class="dt">features=</span>genes.to.use,<span class="dt">only.pos=</span><span class="ot">TRUE</span>)</span></code></pre></div>
<pre><code>## Calculating cluster pbmc</code></pre>
<pre><code>## Calculating cluster tonsil</code></pre>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1"></a>ligs.recs.use &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(markers<span class="op">$</span>gene)</span>
<span id="cb33-2"><a href="#cb33-2"></a><span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(ligs.recs.use)</span></code></pre></div>
<pre><code>## [1] 61</code></pre>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1"></a><span class="co">#Yields 61 ligands and receptors to evaluate</span></span>
<span id="cb35-2"><a href="#cb35-2"></a></span>
<span id="cb35-3"><a href="#cb35-3"></a><span class="co">#Filter ramilowski pairs</span></span>
<span id="cb35-4"><a href="#cb35-4"></a></span>
<span id="cb35-5"><a href="#cb35-5"></a>interactions.forward1 &lt;-<span class="st"> </span>ramilowski_pairs[<span class="kw"><a href="https://rdrr.io/r/base/character.html">as.character</a></span>(ramilowski_pairs<span class="op">$</span>ligand) <span class="op">%in%</span><span class="st"> </span>ligs.recs.use,]</span>
<span id="cb35-6"><a href="#cb35-6"></a>interactions.forward2 &lt;-<span class="st"> </span>ramilowski_pairs[<span class="kw"><a href="https://rdrr.io/r/base/character.html">as.character</a></span>(ramilowski_pairs<span class="op">$</span>receptor) <span class="op">%in%</span><span class="st"> </span>ligs.recs.use,]</span>
<span id="cb35-7"><a href="#cb35-7"></a>interact.for &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span>(interactions.forward1,interactions.forward2)</span>
<span id="cb35-8"><a href="#cb35-8"></a><span class="kw"><a href="https://rdrr.io/r/base/dim.html">dim</a></span>(interact.for)</span></code></pre></div>
<pre><code>## [1] 241   3</code></pre>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1"></a><span class="co">#Yields 241 ligand and receptor interactions to evaluate</span></span>
<span id="cb37-2"><a href="#cb37-2"></a></span>
<span id="cb37-3"><a href="#cb37-3"></a><span class="co">#Create data for celltalker</span></span>
<span id="cb37-4"><a href="#cb37-4"></a></span>
<span id="cb37-5"><a href="#cb37-5"></a>expr.mat &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/Seurat/man/GetAssayData.html">GetAssayData</a></span>(ser.obj,<span class="dt">slot=</span><span class="st">"counts"</span>)</span>
<span id="cb37-6"><a href="#cb37-6"></a>defined.clusters &lt;-<span class="st"> </span>ser.obj<span class="op">@</span>meta.data<span class="op">$</span>cell.types</span>
<span id="cb37-7"><a href="#cb37-7"></a>defined.groups &lt;-<span class="st"> </span>ser.obj<span class="op">@</span>meta.data<span class="op">$</span>sample.type</span>
<span id="cb37-8"><a href="#cb37-8"></a>defined.replicates &lt;-<span class="st"> </span>ser.obj<span class="op">@</span>meta.data<span class="op">$</span>sample.id</span>
<span id="cb37-9"><a href="#cb37-9"></a></span>
<span id="cb37-10"><a href="#cb37-10"></a>reshaped.matrices &lt;-<span class="st"> </span><span class="kw"><a href="../reference/reshape_matrices.html">reshape_matrices</a></span>(<span class="dt">count.matrix=</span>expr.mat,<span class="dt">clusters=</span>defined.clusters,<span class="dt">groups=</span>defined.groups,<span class="dt">replicates=</span>defined.replicates,<span class="dt">ligands.and.receptors=</span>interact.for)</span>
<span id="cb37-11"><a href="#cb37-11"></a></span>
<span id="cb37-12"><a href="#cb37-12"></a><span class="co">#Check out the hierarchy of the tibble</span></span>
<span id="cb37-13"><a href="#cb37-13"></a>reshaped.matrices</span></code></pre></div>
<pre><code>## # A tibble: 2 x 2
##   group  samples         
##   &lt;chr&gt;  &lt;list&gt;          
## 1 pbmc   &lt;tibble [3 × 2]&gt;
## 2 tonsil &lt;tibble [3 × 2]&gt;</code></pre>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1"></a><span class="kw">unnest</span>(reshaped.matrices,<span class="dt">cols=</span><span class="st">"samples"</span>)</span></code></pre></div>
<pre><code>## # A tibble: 6 x 3
##   group  sample   expr.matrices   
##   &lt;chr&gt;  &lt;chr&gt;    &lt;list&gt;          
## 1 pbmc   pbmc_1   &lt;named list [8]&gt;
## 2 pbmc   pbmc_2   &lt;named list [8]&gt;
## 3 pbmc   pbmc_3   &lt;named list [8]&gt;
## 4 tonsil tonsil_1 &lt;named list [8]&gt;
## 5 tonsil tonsil_2 &lt;named list [8]&gt;
## 6 tonsil tonsil_3 &lt;named list [8]&gt;</code></pre>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1"></a><span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="kw">pull</span>(<span class="kw">unnest</span>(reshaped.matrices,<span class="dt">cols=</span><span class="st">"samples"</span>))[[<span class="dv">1</span>]])</span></code></pre></div>
<pre><code>## [1] "b.cells"        "cd14.monocytes" "cd16.monocytes" "cd4.tconv"     
## [5] "cd8.cells"      "nk.cells"       "pdc"            "plasma.cells"</code></pre>
<p>What we have done in this initial step is to separate our overall expression matrix into a series of individual expression matrices from each sample. An overview of the resulting tibble is as follows:</p>
<p>2-column tibble containing group and samples - “samples” column from the top-level contains 3-column tibble - “expr.matrices” from above is a list of count matrices from each cluster of an individual replicate (in this case, clusters are the cell types)</p>
<p>Next, we can create a tibble of consistently expressed ligands and receptors for each group using the create_lig_rec_tib function.</p>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1"></a>consistent.lig.recs &lt;-<span class="st"> </span><span class="kw"><a href="../reference/create_lig_rec_tib.html">create_lig_rec_tib</a></span>(<span class="dt">exp.tib=</span>reshaped.matrices,<span class="dt">clusters=</span>defined.clusters,<span class="dt">groups=</span>defined.groups,<span class="dt">replicates=</span>defined.replicates,<span class="dt">cells.reqd=</span><span class="dv">10</span>,<span class="dt">freq.pos.reqd=</span><span class="fl">0.5</span>,<span class="dt">ligands.and.receptors=</span>interact.for)</span>
<span id="cb43-2"><a href="#cb43-2"></a></span>
<span id="cb43-3"><a href="#cb43-3"></a>consistent.lig.recs</span></code></pre></div>
<pre><code>## # A tibble: 2 x 2
##   group  lig.rec.exp     
##   &lt;chr&gt;  &lt;list&gt;          
## 1 pbmc   &lt;tibble [8 × 2]&gt;
## 2 tonsil &lt;tibble [8 × 2]&gt;</code></pre>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1"></a><span class="kw">unnest</span>(consistent.lig.recs[<span class="dv">1</span>,<span class="dv">2</span>],<span class="dt">cols=</span><span class="st">"lig.rec.exp"</span>)</span></code></pre></div>
<pre><code>## # A tibble: 8 x 2
##   cluster.id     ligands.and.receptors
##   &lt;chr&gt;          &lt;named list&gt;         
## 1 b.cells        &lt;named list [2]&gt;     
## 2 cd14.monocytes &lt;named list [2]&gt;     
## 3 cd16.monocytes &lt;named list [2]&gt;     
## 4 cd4.tconv      &lt;named list [2]&gt;     
## 5 cd8.cells      &lt;named list [2]&gt;     
## 6 nk.cells       &lt;named list [2]&gt;     
## 7 pdc            &lt;named list [2]&gt;     
## 8 plasma.cells   &lt;named list [2]&gt;</code></pre>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1"></a><span class="kw">pull</span>(<span class="kw">unnest</span>(consistent.lig.recs[<span class="dv">1</span>,<span class="dv">2</span>],<span class="dt">cols=</span><span class="st">"lig.rec.exp"</span>)[<span class="dv">1</span>,<span class="dv">2</span>])[[<span class="dv">1</span>]]</span></code></pre></div>
<pre><code>## $ligands
##  [1] "S100A9"   "S100A8"   "IL1B"     "FN1"      "BTLA"     "SPON2"   
##  [7] "LRPAP1"   "VCAN"     "CD14"     "LY86"     "HLA-G"    "HLA-A"   
## [13] "HLA-E"    "HLA-B"    "LTB"      "HSPA1A"   "CD24"     "NAMPT"   
## [19] "TIMP1"    "CD40LG"   "ADAM28"   "PNOC"     "IL7"      "ANXA1"   
## [25] "SEMA4D"   "VIM"      "PSAP"     "LYZ"      "SELPLG"   "HMGB1"   
## [31] "TNFSF13B" "GZMB"     "CALM1"    "SERPINA1" "HSP90AA1" "B2M"     
## [37] "PKM"      "IL16"     "CCL5"     "CCL3"     "ICAM2"    "CD70"    
## [43] "ICAM1"    "ICAM3"    "TGFB1"    "FLT3LG"   "APP"     
## 
## $receptors
##  [1] "CSF3R"     "TGFBR3"    "KCNA3"     "CD53"      "CD1A"      "CD247"    
##  [7] "SELL"      "CXCR4"     "ITGA4"     "GRM7"      "TGFBR2"    "CCR5"     
## [13] "TFRC"      "TLR1"      "IL7R"      "CD180"     "ADRB2"     "CD74"     
## [19] "HMMR"      "HLA-F"     "KCNQ5"     "IGF2R"     "CCR6"      "CD36"     
## [25] "CXCR3"     "PGRMC1"    "CD72"      "TGFBR1"    "ABCA1"     "IFITM1"   
## [31] "CD81"      "KCNQ1"     "CD44"      "CD82"      "IL10RA"    "CD3D"     
## [37] "CD3G"      "CXCR5"     "SORL1"     "APLP2"     "ITGB1"     "FAS"      
## [43] "CD27"      "CD4"       "KLRG1"     "CLEC2B"    "KLRD1"     "KLRC1"    
## [49] "PDE1B"     "CD63"      "TNFRSF17"  "CD19"      "ITGAL"     "ITGAM"    
## [55] "TNFRSF13B" "ERBB2"     "PTPRA"     "CD40"      "OPRL1"     "INSR"     
## [61] "TYROBP"    "CD79A"     "KCNN4"     "FPR2"      "LILRB2"    "LILRB1"   
## [67] "KIR2DL3"   "KIR2DL1"   "KIR3DL2"   "TNFRSF13C" "CELSR1"    "ITGB2"</code></pre>
<p>We can see that we have now recovered a list of consistently expressed ligands and receptors from each cluster in each experimental group, based on the criteria specified above (i.e. at least 10 cells in each cluster express the ligand/receptor and at least 50% of replicates in a group express the ligand/receptor). Functionally, since we only have 2 replicates in each group here, we require all replicate to express the ligand/receptor to count it as consistently expressed.</p>
</div>
<div id="determine-putative-ligandreceptor-pairs" class="section level1">
<h1 class="hasAnchor">
<a href="#determine-putative-ligandreceptor-pairs" class="anchor"></a>Determine putative ligand/receptor pairs</h1>
<p>Now that we have consistently expressed ligands and receptors, we will look for interactions (as defined in ramilowski_pairs$pair) within and between clusters in a given group. The parameter “freq.group.in.cluster” allows us to only examine clusters that consist of (in this case) &gt;5% of cells from a given group for interactions.</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1"></a>put.int &lt;-<span class="st"> </span><span class="kw"><a href="../reference/putative_interactions.html">putative_interactions</a></span>(<span class="dt">ligand.receptor.tibble=</span>consistent.lig.recs,<span class="dt">clusters=</span>defined.clusters,<span class="dt">groups=</span>defined.groups,<span class="dt">freq.group.in.cluster=</span><span class="fl">0.05</span>,<span class="dt">ligands.and.receptors=</span>interact.for)</span></code></pre></div>
<pre><code>## Warning: `cols` is now required.
## Please use `cols = c(lig.rec.exp)`

## Warning: `cols` is now required.
## Please use `cols = c(lig.rec.exp)`</code></pre>
<p>Here, the nested tibble contains the sample groups and a list of ligand/receptor pairs for each group along with the clusters that particpate in the ligand/receptor interactions.</p>
</div>
<div id="identifying-and-visualizing-unique-ligandreceptor-pairs-in-a-group" class="section level1">
<h1 class="hasAnchor">
<a href="#identifying-and-visualizing-unique-ligandreceptor-pairs-in-a-group" class="anchor"></a>Identifying and visualizing unique ligand/receptor pairs in a group</h1>
<p>Now that we have our lists of ligand/receptor interactions, we can investigate which one are unique to each group (using the unique_interactions function) and plot the results (using the circos_plot function) to visualize the differences between groups.</p>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1"></a><span class="co">#Identify unique ligand/receptor interactions present in each sample</span></span>
<span id="cb51-2"><a href="#cb51-2"></a>unique.ints &lt;-<span class="st"> </span><span class="kw"><a href="../reference/unique_interactions.html">unique_interactions</a></span>(put.int,<span class="dt">group1=</span><span class="st">"pbmc"</span>,<span class="dt">group2=</span><span class="st">"tonsil"</span>,interact.for)</span>
<span id="cb51-3"><a href="#cb51-3"></a></span>
<span id="cb51-4"><a href="#cb51-4"></a><span class="co">#Get data to plot circos for PBMC</span></span>
<span id="cb51-5"><a href="#cb51-5"></a>pbmc.to.plot &lt;-<span class="st"> </span><span class="kw">pull</span>(unique.ints[<span class="dv">1</span>,<span class="dv">2</span>])[[<span class="dv">1</span>]]</span>
<span id="cb51-6"><a href="#cb51-6"></a>for.circos.pbmc &lt;-<span class="st"> </span><span class="kw">pull</span>(put.int[<span class="dv">1</span>,<span class="dv">2</span>])[[<span class="dv">1</span>]][pbmc.to.plot]</span>
<span id="cb51-7"><a href="#cb51-7"></a></span>
<span id="cb51-8"><a href="#cb51-8"></a><span class="kw"><a href="../reference/circos_plot.html">circos_plot</a></span>(<span class="dt">interactions=</span>for.circos.pbmc,<span class="dt">clusters=</span>defined.clusters)</span></code></pre></div>
<pre><code>## Note: 1 point is out of plotting region in sector 'b.cells', track '1'.</code></pre>
<pre><code>## Note: 1 point is out of plotting region in sector 'cd14.monocytes',
## track '1'.</code></pre>
<pre><code>## Note: 1 point is out of plotting region in sector 'cd16.monocytes',
## track '1'.</code></pre>
<pre><code>## Note: 1 point is out of plotting region in sector 'cd4.tconv', track
## '1'.</code></pre>
<pre><code>## Note: 1 point is out of plotting region in sector 'cd8.cells', track
## '1'.</code></pre>
<pre><code>## Note: 1 point is out of plotting region in sector 'nk.cells', track
## '1'.</code></pre>
<pre><code>## Note: 1 point is out of plotting region in sector 'pdc', track '1'.</code></pre>
<pre><code>## Note: 1 point is out of plotting region in sector 'plasma.cells', track
## '1'.</code></pre>
<p><img src="celltalker_files/figure-html/unnamed-chunk-9-1.png" width="960" style="display: block; margin: auto;"></p>
<pre><code>## [1] "Plotted interactions"</code></pre>
<div class="sourceCode" id="cb61"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1"></a><span class="co">#Get data to plot circos for tonsil</span></span>
<span id="cb61-2"><a href="#cb61-2"></a>tonsil.to.plot &lt;-<span class="st"> </span><span class="kw">pull</span>(unique.ints[<span class="dv">2</span>,<span class="dv">2</span>])[[<span class="dv">1</span>]]</span>
<span id="cb61-3"><a href="#cb61-3"></a>for.circos.tonsil &lt;-<span class="st"> </span><span class="kw">pull</span>(put.int[<span class="dv">2</span>,<span class="dv">2</span>])[[<span class="dv">1</span>]][tonsil.to.plot]</span>
<span id="cb61-4"><a href="#cb61-4"></a></span>
<span id="cb61-5"><a href="#cb61-5"></a><span class="kw"><a href="../reference/circos_plot.html">circos_plot</a></span>(<span class="dt">interactions=</span>for.circos.tonsil,<span class="dt">clusters=</span>defined.clusters)</span></code></pre></div>
<pre><code>## Note: 1 point is out of plotting region in sector 'b.cells', track '1'.</code></pre>
<pre><code>## Note: 1 point is out of plotting region in sector 'cd14.monocytes',
## track '1'.</code></pre>
<pre><code>## Note: 1 point is out of plotting region in sector 'cd16.monocytes',
## track '1'.</code></pre>
<pre><code>## Note: 1 point is out of plotting region in sector 'cd4.tconv', track
## '1'.</code></pre>
<pre><code>## Note: 1 point is out of plotting region in sector 'cd8.cells', track
## '1'.</code></pre>
<pre><code>## Note: 1 point is out of plotting region in sector 'nk.cells', track
## '1'.</code></pre>
<pre><code>## Note: 1 point is out of plotting region in sector 'pdc', track '1'.</code></pre>
<pre><code>## Note: 1 point is out of plotting region in sector 'plasma.cells', track
## '1'.</code></pre>
<p><img src="celltalker_files/figure-html/unnamed-chunk-9-2.png" width="960" style="display: block; margin: auto;"></p>
<pre><code>## [1] "Plotted interactions"</code></pre>
</div>
<div id="summary" class="section level1">
<h1 class="hasAnchor">
<a href="#summary" class="anchor"></a>Summary</h1>
<p>celltalker allows a user to infer ligand/receptor interactions. This vignette has walked through how to identify and visualize unique interactions, staring with raw count data and metadata. Future developments will focus on differential networks of cell-cell communication for given ligand/receptor pairs. Stay tuned for future updates. Until then, have fun investigating how your cells “talk” to each other!</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li><a href="#vignette-overview">Vignette overview</a></li>
      <li><a href="#clustering-data-with-seurat">Clustering data with Seurat</a></li>
      <li><a href="#celltalker">celltalker</a></li>
      <li><a href="#consistently-expressed-ligands-and-receptors">Consistently expressed ligands and receptors</a></li>
      <li><a href="#determine-putative-ligandreceptor-pairs">Determine putative ligand/receptor pairs</a></li>
      <li><a href="#identifying-and-visualizing-unique-ligandreceptor-pairs-in-a-group">Identifying and visualizing unique ligand/receptor pairs in a group</a></li>
      <li><a href="#summary">Summary</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Anthony Cillo.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
